js/modules/k6/experimental/tracing/tracing.go

```
@@ -0,0 +1,286 @@
+	// Propagation is the propagation format to use for the tracer.
+	Propagator string `js:"propagator"`
+
+	// Sampling is the sampling rate to use for the tracer.
+	Sampling *float64 `js:"sampling"`
+
+	// Baggage is a map of baggage items to add to the tracer.
+	Baggage map[string]string `js:"baggage"`
+}
+
+func (i *instrumentationOptions) validate() error {
+	var (
+		isW3C    = i.Propagator == W3CPropagatorName
+		isB3     = i.Propagator == B3PropagatorName
+		isJaeger = i.Propagator == JaegerPropagatorName
+	)
+	if !isW3C && !isB3 && !isJaeger {
+		return fmt.Errorf("unknown propagator: %s", i.Propagator)
+	}
+
+	if i.Sampling != nil {
+		return fmt.Errorf("sampling is not yet supported")
+	}
+
+	if i.Baggage != nil {
+		return fmt.Errorf("baggage is not yet supported")
+	}
+
+	return nil
+}
+
+// instrumentHTTPMethod returns a new function that wraps the original http
+// method, adding tracing headers to the request.
+//
+// The function takes the HTTP method name as argument, as well as the exported
+// method itself, as a goja.Callable.
+//
+// The produced resulting function is ready to be injected back into the http
+// module, in place of the original method.
+func (t *Tracing) instrumentHTTPMethod(methodName k6HTTPMethodName, methodFn goja.Callable) goja.Callable {
+	return func(this goja.Value, args ...goja.Value) (goja.Value, error) {
+		rt := t.vu.Runtime()
+
+		// Ensure the arguments have a params object, in which
+		// we can add the tracing headers at a later point in time.
+		args, params, err := t.getOrCreateParams(methodName, args...)
+		if err != nil {
+			common.Throw(rt, fmt.Errorf("failed to normalize the params argument; reason: %w", err))
+		}
+
+		// Ensure that the params object contains a headers object.
+		// Create it if it doesn't.
+		headers, err := t.getOrCreateHeaders(params)
+		if err != nil {
+			common.Throw(rt, fmt.Errorf("failed to normalize params argument headers property; reason: %w", err))
+		}
```

This `headers` value is first used in line `168`. Perhaps it makes sense to move this call a bit closer to that line. You can read more about this in Clean Code, Chapter 5 for example.

---

In other scenarios, I would agree, and my C++ years tend to lead me to prefer code locality indeed. However, in this example, the choice I made was to prefer failing early. The trace id generation issues syscalls (reading from the os random source) to produce randomness which has a cost, thus I wanted to make sure that I would produce a trace id only if I was sure I'd have somewhere to put it.

---

Got it 👍